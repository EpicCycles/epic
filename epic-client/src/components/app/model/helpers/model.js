import {
    findObjectWithKey,
    generateRandomCode,
    isItAnObject,
    removeKey,
    updateObject,
    updateObjectInArray
} from "../../../../helpers/utils";
import {NEW_ELEMENT_ID} from "../../../../helpers/constants";

export const eliminateReadOnlyFields = (fieldList) => {
    return fieldList.filter(field => (! field.readOnly));
};

export const checkForChanges = (fieldList, existingObject, newValues) => {
    return eliminateReadOnlyFields(fieldList).some(field => {
        return existingObject[field.fieldName] !== newValues[field.fieldName];
    })
};

export const addFieldToState = (initialState, fieldList, fieldName, input) => {
    let finalState = initialState;
    fieldList.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            finalState = applyFieldValueToModel(initialState, field, input);
            return true;
        }
        return false;
    });
    return finalState;
};


export const getModelKey = modelInstance => {
    if (!modelInstance) return NEW_ELEMENT_ID;
    if (modelInstance.id) return modelInstance.id;
    if (modelInstance.dummyKey) return modelInstance.dummyKey;
    return NEW_ELEMENT_ID;
};

export const isModelValid = (modelInstance) => {
    return !(modelInstance.error || isItAnObject(modelInstance.error_detail));
};

export const applyFieldValueToModel = (modelInstance, field, value) => {
    let updatedModelInstance = updateObject(modelInstance);
    updatedModelInstance[field.fieldName] = value;
    updatedModelInstance.changed = true;
    if (!updatedModelInstance.error_detail) updatedModelInstance.error_detail = {};
    if (field.required && !value) {
        updatedModelInstance.error_detail[field.fieldName] = field.error;
        return updatedModelInstance;
    } else if (field.validator) {
        const error = field.validator(value, modelInstance);
        if (error) {
            updatedModelInstance.error_detail[field.fieldName] = error;
            return updatedModelInstance;
        }
    }
    updatedModelInstance.error_detail = removeKey(updatedModelInstance.error_detail, field.fieldName);
    return updatedModelInstance;
};

export const getAttribute = (modelFields, fieldName) => {
    let attribute;
    modelFields.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            attribute = field.fieldName;
            return true;
        }
        return false;
    });
    return attribute;
};

export const getField = (modelFields, fieldName) => {
    let modelField;
    modelFields.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            modelField = field;
            return true;
        }
        return false;
    });
    return modelField;
};

export const updateModelChanges = (modelInstance, attribute, fieldValue) => {
    let changes = updateObject(modelInstance.changes);
    if (fieldValue && (modelInstance[attribute] !== fieldValue)) {
        changes[attribute] = fieldValue;
    } else {
        changes = removeKey(changes, attribute);
    }
    return updateObject(modelInstance, { changes })
};
export const updateModelWithChanges = (modelInstance, modelFields, fieldName, fieldValue) => {
    const attribute = getAttribute(modelFields, fieldName);
    if (attribute) {
        return updateModelChanges(modelInstance, attribute, fieldValue)
    } else {
        console.error("attribute not found for", modelFields, fieldName)
    }
    return modelInstance;
};
export const updateModelArrayWithChanges = (modelArray, modelFields, fieldName, fieldValue, componentKey) => {
    const attribute = getAttribute(modelFields, fieldName);
    if (attribute) {
        let modelInstance = findObjectWithKey(modelArray, componentKey);
        let currentChanges = modelInstance.changes || {};
        if (fieldValue && (modelInstance[attribute] !== fieldValue)) {
            currentChanges[attribute] = fieldValue;
        } else {
            currentChanges = removeKey(currentChanges, attribute);
        }
        modelInstance.changes = currentChanges;
        return updateObjectInArray(modelArray, modelInstance, componentKey)
    } else {
        console.error("attribute not found for", modelFields, fieldName)
    }
    return modelArray;
};
export const updateModel = (model, modelFields, fieldName, fieldValue, componentKey) => {
    const modelField = getField(modelFields, fieldName);
    if (modelField) return applyFieldValueToModel(model, modelField, fieldValue);
    return model;
};

export const displayModelErrorSummary = (model, modelFields) => {
    let displayErrors = model.error;
    const fieldErrors = isItAnObject(model.error_detail) ? model.error_detail : {};
    for (var property in fieldErrors) {
        const field = getField(modelFields, property);
        if (!displayErrors) {
            displayErrors = "";
        } else {
            displayErrors += '<br>';
        }
        displayErrors += field.header + ": " + fieldErrors[property].join(" ");
    }

    return displayErrors;
};

export const getNameForValue = (value, sourceArray) => {
    if (!value) return undefined;
    let name = value;

    sourceArray.some(pair => {
        if (pair.value === value) {
            name = pair.name;
            return true;
        }
        return false;
    });
    return name;
};

export const hasErrors = (model) => {
    if (model.error && model.error.length > 0) return true;
    if (isItAnObject(model.error_detail)) return true;
    return false;
};

export const addErrorDetail = (error_detail = {}, field, error) => {
      let updated_error_detail = updateObject(error_detail);
  if (!Array.isArray(updated_error_detail[field])) {
         updated_error_detail[field] = [];
  }
  updated_error_detail[field].push(error);
  return updated_error_detail;
};

export const createNewModelInstance = () => {
    return {dummyKey: generateRandomCode()};
};
export const matchesModel = (persistedModel, modelFields, modelToCheck) => {
     return eliminateReadOnlyFields(modelFields).every(field => {
        return persistedModel[field.fieldName] === modelToCheck[field.fieldName];
    })
};