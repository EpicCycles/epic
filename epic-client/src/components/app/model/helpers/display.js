export const fixedHeaderClassname = (lockColumn) => {
    if (lockColumn) return "grid-header--fixed-left";
    return "";
};
export const fixedDetailsClassname = (lockColumn) => {
    if (lockColumn) return "grid-item--fixed-left";
    return "";
};
export const gridHeaderClass = (baseClassName = "", fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && (fieldIndex === 0);
    return `${baseClassName} grid-item--header ${fixedHeaderClassname(shouldLock)}`;
};
export const gridItemClass = (baseClassName = "", fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && (fieldIndex === 0);
    return `${baseClassName} grid-item ${fixedDetailsClassname(shouldLock)}`;
};