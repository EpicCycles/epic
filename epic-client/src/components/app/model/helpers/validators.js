import {INVALID_POSTCODE} from "./error";
import {POSTCODE_RULES} from "./constants";
export const validateData = (fieldList, currentValues) => {
    let errors = {};
    fieldList.forEach(field => {
        if (field.required) {
            if (!currentValues[field.fieldName]) {
                errors[field.fieldName] = field.error;
            }
        }
    });
    return errors;
};

export const validateEmailFormat = (email) => {
    const emailPattern = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return emailPattern.test(email);
};

export const validatePostcodeAndReturnError = (postcode, model = {}) => {
    let error;
    const postcodeRule = POSTCODE_RULES.filter(rule => rule.countryCode === model.country);
    if (postcodeRule[0]) {
        const OK = RegExp(postcodeRule[0].regex).exec(postcode);
        if (!OK) error = INVALID_POSTCODE + postcodeRule[0].display;
    }
    return error;
};