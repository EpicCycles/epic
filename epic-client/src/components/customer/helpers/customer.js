// builds a string for the customer
export const buildCustomerString = (customer) => {
   let displayArray = [];
   if (customer.first_name) displayArray.push(customer.first_name);
   if (customer.last_name) displayArray.push(customer.last_name);
   if (customer.email) displayArray.push(`(${customer.email})`);
   return displayArray.join(' ');
};