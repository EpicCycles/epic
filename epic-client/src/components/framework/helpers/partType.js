import {updateObject} from "../../../helpers/utils";
import {PART_TYPE_NAME_MISSING} from "../../app/model/helpers/error";
import {NEW_ELEMENT_ID} from "../../../helpers/constants";

export const getPartTypeName = (partTypeId, sections) => {
    if (!partTypeId) return undefined;
    const partType = getPartType(partTypeId, sections);
    if (partType) return partType.name;
    return 'Unknown Part Type';
};
export const getPartType = (partTypeId, sections) => {
    if (!partTypeId) return undefined;
    let partType;
    sections.some(section => {
        return section.partTypes.some(aPartType => {
            if (aPartType.id === partTypeId) {
                partType = aPartType;
                return true;
            }
            return false;
        });
    });
    return partType;
};
export const getPartTypeAttributeFields = (partTypeId, sections) => {
    const partType = getPartType(partTypeId, sections);
    if (partType) return partType.attributes;
    return [];
};
export const processPartTypeValueChanges = (partType, componentKey, fieldName, input) => {
    const updatedPartType = updateObject(partType);
    if (fieldName.startsWith('name')) updatedPartType.name = input;
    if (!updatedPartType.name) {
        updatedPartType.error = true;
        updatedPartType.error_detail = PART_TYPE_NAME_MISSING;
    } else {
        updatedPartType.error = false;
        updatedPartType.error_detail = "";
    }
    if (fieldName.startsWith('description')) updatedPartType.description = input;
    if (fieldName.startsWith('can_be_substituted')) updatedPartType.can_be_substituted = input;
    if (fieldName.startsWith('can_be_omitted')) updatedPartType.can_be_omitted = input;
    if (fieldName.startsWith('customer_visible')) updatedPartType.customer_visible = input;
    if (fieldName.startsWith('attributes')) updatedPartType.attributes = input;
    if (fieldName.startsWith('synonyms')) updatedPartType.synonyms = input;
    if (fieldName.startsWith('detail')) updatedPartType._detail = input;
    if (componentKey === NEW_ELEMENT_ID) updatedPartType.dummyKey = NEW_ELEMENT_ID;

    updatedPartType.changed = true;
    return updatedPartType;
};
export const doesFieldMatchPartType = (partType, fieldName) => {
    const fieldNameLower = fieldName.toLowerCase();
    if (partType.name.toLowerCase() === fieldNameLower) {
        return true;
    } else {
        return partType.synonyms.some(synonym => {
            return (synonym.name.toLowerCase() === fieldNameLower);
        });
    }
};