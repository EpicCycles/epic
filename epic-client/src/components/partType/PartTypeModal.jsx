import React from "react";
import ReactModal from 'react-modal';

import * as PropTypes from "prop-types";
import PartTypeData from "./PartTypeData";
import {Icon} from "semantic-ui-react";
import {SECTION_MISSING} from "../../helpers/error";
import SelectInput from "../../common/SelectInput";
import {processPartTypeValueChanges} from "../../helpers/framework";

class PartTypeModal extends React.Component {
    constructor(props) {
        super();
        this.state = this.deriveStateFromProps(props);
    };
    deriveStateFromProps = (props) => {
        const {sections, partType} = props;
        const sectionOptions = sections.map(section => {
            return { name: section.name, value: section.id.toString() }
        });
        return {
            sectionOptions,
            partType: partType || {},
            mode: (partType && partType.id) ? "Edit" : "New",
        }
    };
    handlePartTypeValueChange = (fieldName, input) => {
        const updatedPartType = processPartTypeValueChanges(this.state.partType, this.state.componentKey, fieldName, input);
        if (fieldName.startsWith("includeInSection")) updatedPartType.includeInSection = input;
        if (!updatedPartType.includeInSection) {
            if (updatedPartType.error) {
                updatedPartType.error_detail = SECTION_MISSING + '.' + updatedPartType.error_detail;
            } else {
                updatedPartType.error = true;
                updatedPartType.error_detail = SECTION_MISSING;
            }
        }
        this.setState({ partType: updatedPartType });
    };
    onClickReset = () => {
        this.setState(this.deriveStateFromProps(this.props));
    };
    onAfterOpen = () => {
        this.setState(this.deriveStateFromProps(this.props));
    };

    saveOrCreatePartType = () => {
        this.props.savePartType(this.state.partType);
        this.props.closePartTypeModal();
    };
    deleteOrRemovePartType = () => {
        if (this.state.componentKey) {
            this.props.deletePartType(this.state.partType);
        }
        this.props.closePartTypeModal();
    };

    render() {
        const { partTypeModalOpen, componentKey, deletePartType, closePartTypeModal } = this.props;
        const { partType, mode, sectionOptions } = this.state;
        return <ReactModal
            isOpen={partTypeModalOpen}
            contentLabel={`${mode} Part Type`}
            className="Modal PartTypeModal"
            onAfterOpen={this.onAfterOpen}
        >
            <div style={{ width: "100%", textAlign: "right" }}>
                <Icon
                    name="remove"
                    circular
                    link
                    onClick={closePartTypeModal}
                />
            </div>
            {partTypeModalOpen && <div style={{ width: "100%", textAlign: "left" }}>
                <h2>{mode} Part Type</h2>
                {partType.error && <div className="red">{partType.error_detail}</div>}
                <div>
                    <SelectInput
                        className=""
                        title="Select Section"
                        label="Section"
                        fieldName="includeInSection"
                        onChange={this.handlePartTypeValueChange}
                        value={partType.includeInSection  && partType.includeInSection.toString()}
                        options={sectionOptions}
                        isEmptyAllowed={true}
                    />
                </div>
                <PartTypeData
                    partType={partType}
                    componentKey={componentKey}
                    handlePartTypeValueChange={this.handlePartTypeValueChange}
                />
            </div>}
            <div style={{ width: "100%", textAlign: "right" }}>
                {partType.changed &&
                <Icon id={`reset-partType`} name="undo"
                      onClick={this.onClickReset}
                      title="Reset PartType details"
                />
                }
                {(partType.changed && (!partType.error)) &&
                <Icon id={`accept-partType`} name="check"
                      onClick={this.saveOrCreatePartType}
                      title="Confirm PartType Change"/>
                }
                {(deletePartType && (partType.id || partType.changed)) &&
                <Icon id={`delete-partType`} name="trash"
                      onClick={this.deleteOrRemovePartType}
                      title="Delete PartType"/>
                }
            </div>
        </ReactModal>;
    }
}

PartTypeModal.propTypes = {
    partTypeModalOpen: PropTypes.any,
    partType: PropTypes.any,
    componentKey: PropTypes.any,
    savePartType: PropTypes.any,
    sections: PropTypes.any,
    closePartTypeModal: PropTypes.func,
    deletePartType: PropTypes.any,
};

export default PartTypeModal;