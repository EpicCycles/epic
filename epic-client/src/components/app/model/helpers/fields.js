import {
    ADDRESS_MISSING,
    BRAND_MISSING, BRAND_NAME_MISSING,
    BUNDLE_NAME_MISSING, COUNTRY_MISSING,
    FRAME_NAME_MISSING,
    MODEL_NAME_MISSING,
    PART_MISSING,
    PART_NAME_MISSING,
    PART_TYPE_MISSING,
    PRODUCTS_MISSING,
    SUPPLIER_MISSING
} from "./error";
import {validatePostcodeAndReturnError} from "./validators";

// TODO customer fields including email with email validator
export const ADDRESS1 = "address1";
export const ADDRESS2 = "address2";
export const ADDRESS3 = "address3";
export const ADDRESS4 = "address4";
export const COUNTRY = "country";
export const POSTCODE = "postcode";
export const BRAND = "brand";
export const SUPPLIER = "supplier";
export const PART = "part";
export const LINK = "link";
export const BUNDLE = "bundle_name";
export const PRODUCT_CODE = "product_code";
export const PART_TYPE = "partType";
export const BRAND_NAME = "brand_name";
export const BIKE_BRAND = "bike_brand";
export const FRAME_NAME = "frame_name";
export const MODEL_NAME = "model_name";
export const PART_NAME = "part_name";
export const DESCRIPTION = "description";
export const COLOURS = "colours";
export const SELL_PRICE = "rrp";
export const EPIC_PRICE = "epic_price";
export const TICKET_PRICE = "ticket_price";
export const TRADE_PRICE = "trade_price";
export const CLUB_PRICE = "club_price";
export const FITTED_PRICE = "fitted_price";
export const TRADE_IN_PRICE = "trade_in_value";
export const SIZES = "sizes";
export const STANDARD = "standard";
export const STOCKED = "stocked";
export const PRODUCTS = "products";
export const TEXT = "text";
export const NUMBER = "number";
export const CURRENCY = "currency";
export const TEXT_AREA = "textArea";
export const CHECKBOX = "checkbox";

// TODO add COUNTRY type and fields for address
// TODO consider having a function passed for validator
export const ADDRESS1_FIELD = {
    fieldName: ADDRESS1,
    type: TEXT,
    length: 100,
    header: "Address",
    required: true,
    error: ADDRESS_MISSING
};
export const ADDRESS2_FIELD = {
    fieldName: ADDRESS2,
    type: TEXT,
    length: 100,
    header: "Line 2",
};
export const ADDRESS3_FIELD = {
    fieldName: ADDRESS3,
    type: TEXT,
    length: 100,
    header: "Line 3",
};
export const ADDRESS4_FIELD = {
    fieldName: ADDRESS4,
    type: TEXT,
    length: 100,
    header: "Line 4",
};
export const COUNTRY_FIELD = {
    fieldName: COUNTRY,
    header: "Country",
    type: COUNTRY,
    required: true,
    error: COUNTRY_MISSING
};
export const POSTCODE_FIELD = {
    fieldName: POSTCODE,
    type: TEXT,
    length: 20,
    header: "Postcode",
    validator: validatePostcodeAndReturnError,
    validatorAdditionalFields: [COUNTRY]
};
export const BRAND_FIELD = {
    fieldName: BRAND,
    header: "Brand",
    synonyms: [BRAND, 'manufacturer'],
    required: true,
    error: BRAND_MISSING,
    type: BRAND
};
export const PRODUCTS_FIELD = {
    fieldName: PRODUCTS,
    header: "Products",
    synonyms: [],
    required: true,
    error: PRODUCTS_MISSING,
    type: PRODUCTS
};
export const SUPPLIER_FIELD = {
    fieldName: SUPPLIER,
    header: "Supplier",
    synonyms: [SUPPLIER],
    required: true,
    error: SUPPLIER_MISSING,
    type: SUPPLIER
};
export const SUPPLIER_FIELD_OPTIONAL = {
    fieldName: SUPPLIER,
    header: "Supplier",
    synonyms: [SUPPLIER],
    type: SUPPLIER
};
export const PART_FIELD = {
    fieldName: PART,
    header: "Part",
    synonyms: [PART],
    required: true,
    error: PART_MISSING,
    type: PART
};
export const FRAME_NAME_FIELD = {
    fieldName: FRAME_NAME,
    header: "Frame Name",
    synonyms: [],
    required: true,
    error: FRAME_NAME_MISSING
};
export const LINK_FIELD = {
    fieldName: LINK,
    header: "URL",
    synonyms: [],
};
export const BRAND_NAME_FIELD = {
    fieldName: BRAND_NAME,
    header: "Brand Name",
    synonyms: [],
    required: true,
    error: BRAND_NAME_MISSING
};
export const BUNDLE_NAME_FIELD = {
    fieldName: BUNDLE,
    header: "Bundle Name",
    synonyms: [],
    required: true,
    error: BUNDLE_NAME_MISSING
};
export const MODEL_NAME_FIELD = {
    fieldName: MODEL_NAME,
    header: "Model Name",
    synonyms: [],
    required: true,
    error: MODEL_NAME_MISSING,
    type: TEXT,
    length: 100
};
export const COLOURS_FIELD = {
    fieldName: COLOURS,
    header: "Colours",
    synonyms: [COLOURS, "colour", "colors", "color"],
    type: TEXT,
    length: 100
};
export const DESCRIPTION_FIELD = {
    fieldName: DESCRIPTION,
    header: "Description",
    synonyms: [DESCRIPTION, "desc"],
    type: TEXT_AREA,
    length: 400
};
export const PRODUCT_CODE_FIELD = {
    fieldName: PRODUCT_CODE,
    header: "Product Code",
    synonyms: [PRODUCT_CODE],
    type: TEXT,
    length: 30
};
export const SELL_PRICE_FIELD = {
    fieldName: SELL_PRICE,
    header: "RRP",
    synonyms: ["price", "selling price", "srp", "rrp", "sell price", "retail price"],
    type: CURRENCY,
    length: 10
};
export const EPIC_PRICE_FIELD = {
    fieldName: EPIC_PRICE,
    header: "Epic Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const CLUB_PRICE_FIELD = {
    fieldName: CLUB_PRICE,
    header: "Club Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};

export const FITTED_PRICE_FIELD = {
    fieldName: FITTED_PRICE,
    header: "Fitted Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const TRADE_PRICE_FIELD = {
    fieldName: TRADE_PRICE,
    header: "Trade Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const TICKET_PRICE_FIELD = {
    fieldName: TICKET_PRICE,
    header: "Ticket Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const SIZES_FIELD = {
    fieldName: SIZES,
    header: "Sizes",
    synonyms: [SIZES, "size", "frame sizes", "frame size"],
    type: TEXT,
    length: 100
};
export const PART_TYPE_FIELD = {
    fieldName: PART_TYPE,
    header: "Part Type",
    required: true,
    error: PART_TYPE_MISSING,
    type: PART_TYPE,
};
export const PART_NAME_FIELD = {
    fieldName: PART_NAME,
    header: "Part Name",
    required: true,
    error: PART_NAME_MISSING
};
export const TRADE_IN_FIELD = {
    fieldName: TRADE_IN_PRICE,
    header: "Trade In Â£",
    type: CURRENCY,
    length: 10
};
export const STOCKED_FIELD = {
    fieldName: STOCKED,
    header: "Stocked",
    type: CHECKBOX
};
export const STANDARD_FIELD = {
    fieldName: STANDARD,
    header: "Standard",
    type: CHECKBOX
};
export const BIKE_BRAND_FIELD = {
    fieldName: BIKE_BRAND,
    header: "Bike Brand",
    type: CHECKBOX
};
export const customerAddressFields = [ADDRESS1_FIELD, ADDRESS2_FIELD, ADDRESS3_FIELD, ADDRESS4_FIELD, COUNTRY_FIELD, POSTCODE_FIELD];
export const frameFields = [BRAND, FRAME_NAME];
export const brandFields = [BRAND_NAME_FIELD, BIKE_BRAND_FIELD, LINK_FIELD, SUPPLIER_FIELD_OPTIONAL];
export const bikeFields = [MODEL_NAME_FIELD, DESCRIPTION_FIELD, COLOURS_FIELD, SELL_PRICE_FIELD, EPIC_PRICE_FIELD, CLUB_PRICE_FIELD, SIZES_FIELD];
export const partFields = [
    PART_TYPE_FIELD,
    BRAND_FIELD,
    PART_NAME_FIELD,
    TRADE_IN_FIELD,
    STANDARD_FIELD,
    STOCKED_FIELD
];
export const supplierProductFields = [
    SUPPLIER_FIELD,
    FITTED_PRICE_FIELD,
    TICKET_PRICE_FIELD,
    SELL_PRICE_FIELD,
    TRADE_PRICE_FIELD,
    CLUB_PRICE_FIELD
];
export const bundleFields = [
    BUNDLE_NAME_FIELD,
    PRODUCTS_FIELD,
    FITTED_PRICE_FIELD,
    TICKET_PRICE_FIELD,
];

