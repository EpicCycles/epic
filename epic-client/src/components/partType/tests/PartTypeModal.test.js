import React from 'react';
import PartTypeModal from "../PartTypeModal";
import {PART_TYPE_NAME_MISSING, SECTION_MISSING} from "../../app/model/helpers/error";

const sections = [
    { name: "section 1", id: 1 },
    { name: "section 2", id: 2 }
];
test("displays correctly with a new part", () => {
    const component = shallow(<PartTypeModal
        partTypeModalOpen={true}
        sections={sections}
    />);
    expect(component).toMatchSnapshot();
});
test("displays correctly with a partial new part", () => {
    const partType = {
        shortName: "new one",
        _detail: true
    };
    const component = shallow(<PartTypeModal
        partTypeModalOpen={true}
        sections={sections}
        partType={partType}
    />);
    expect(component).toMatchSnapshot();
});
test("error when no short name", () => {
    const partType = {
        includeInSection:sections[0].id,
        shortName: "has one one",
        id: "1",
        _detail: true
    };
    const component = shallow(<PartTypeModal
        partTypeModalOpen={true}
        sections={sections}
        partType={partType}
    />);
    expect(component).toMatchSnapshot();

    component.instance().handlePartTypeValueChange("shortName_1", "");
    const expectedData = Object.assign({}, partType, {
        shortName: "",
        changed: true,
        error: true,
        error_detail: PART_TYPE_NAME_MISSING
    });
    expect(component.state('partType')).toEqual(expectedData);

});
test("error when no includeInSection", () => {
    const partType = {
        includeInSection:sections[0].id,
        shortName: "has one one",
        id: "1",
        _detail: true
    };
    const component = shallow(<PartTypeModal
        partTypeModalOpen={true}
        sections={sections}
        partType={partType}
    />);
    expect(component).toMatchSnapshot();

    component.instance().handlePartTypeValueChange("includeInSection_1", "");
    const expectedData = Object.assign({}, partType, {
        includeInSection: "",
        changed: true,
        error: true,
        error_detail: SECTION_MISSING
    });
    expect(component.state('partType')).toEqual(expectedData);

});