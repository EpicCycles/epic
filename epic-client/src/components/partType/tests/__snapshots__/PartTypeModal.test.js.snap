// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays correctly with a new part 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="Modal PartTypeModal"
  closeTimeoutMS={0}
  contentLabel="New Part Type"
  isOpen={true}
  onAfterOpen={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div
    style={
      Object {
        "textAlign": "right",
        "width": "100%",
      }
    }
  >
    <Icon
      as="i"
      circular={true}
      link={true}
      name="remove"
    />
  </div>
  <div
    style={
      Object {
        "textAlign": "left",
        "width": "100%",
      }
    }
  >
    <h2>
      New
       Part Type
    </h2>
    <div>
      <SelectInput
        className=""
        error=""
        fieldName="includeInSection"
        isEmptyAllowed={true}
        label="Section"
        multipleSize={1}
        onChange={[Function]}
        options={
          Array [
            Object {
              "name": "section 1",
              "value": "1",
            },
            Object {
              "name": "section 2",
              "value": "2",
            },
          ]
        }
        title="Select Section"
      />
    </div>
    <PartTypeData
      handlePartTypeValueChange={[Function]}
      partType={Object {}}
    />
  </div>
  <div
    style={
      Object {
        "textAlign": "right",
        "width": "100%",
      }
    }
  />
</Modal>
`;

exports[`displays correctly with a partial new part 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="Modal PartTypeModal"
  closeTimeoutMS={0}
  contentLabel="New Part Type"
  isOpen={true}
  onAfterOpen={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div
    style={
      Object {
        "textAlign": "right",
        "width": "100%",
      }
    }
  >
    <Icon
      as="i"
      circular={true}
      link={true}
      name="remove"
    />
  </div>
  <div
    style={
      Object {
        "textAlign": "left",
        "width": "100%",
      }
    }
  >
    <h2>
      New
       Part Type
    </h2>
    <div>
      <SelectInput
        className=""
        error=""
        fieldName="includeInSection"
        isEmptyAllowed={true}
        label="Section"
        multipleSize={1}
        onChange={[Function]}
        options={
          Array [
            Object {
              "name": "section 1",
              "value": "1",
            },
            Object {
              "name": "section 2",
              "value": "2",
            },
          ]
        }
        title="Select Section"
      />
    </div>
    <PartTypeData
      handlePartTypeValueChange={[Function]}
      partType={
        Object {
          "_detail": true,
          "name": "new one",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "textAlign": "right",
        "width": "100%",
      }
    }
  />
</Modal>
`;

exports[`error when no includeInSection 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="Modal PartTypeModal"
  closeTimeoutMS={0}
  contentLabel="Edit Part Type"
  isOpen={true}
  onAfterOpen={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div
    style={
      Object {
        "textAlign": "right",
        "width": "100%",
      }
    }
  >
    <Icon
      as="i"
      circular={true}
      link={true}
      name="remove"
    />
  </div>
  <div
    style={
      Object {
        "textAlign": "left",
        "width": "100%",
      }
    }
  >
    <h2>
      Edit
       Part Type
    </h2>
    <div>
      <SelectInput
        className=""
        error=""
        fieldName="includeInSection"
        isEmptyAllowed={true}
        label="Section"
        multipleSize={1}
        onChange={[Function]}
        options={
          Array [
            Object {
              "name": "section 1",
              "value": "1",
            },
            Object {
              "name": "section 2",
              "value": "2",
            },
          ]
        }
        title="Select Section"
        value="1"
      />
    </div>
    <PartTypeData
      handlePartTypeValueChange={[Function]}
      partType={
        Object {
          "_detail": true,
          "id": "1",
          "includeInSection": 1,
          "name": "has one one",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "textAlign": "right",
        "width": "100%",
      }
    }
  />
</Modal>
`;

exports[`error when no short name 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="Modal PartTypeModal"
  closeTimeoutMS={0}
  contentLabel="Edit Part Type"
  isOpen={true}
  onAfterOpen={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div
    style={
      Object {
        "textAlign": "right",
        "width": "100%",
      }
    }
  >
    <Icon
      as="i"
      circular={true}
      link={true}
      name="remove"
    />
  </div>
  <div
    style={
      Object {
        "textAlign": "left",
        "width": "100%",
      }
    }
  >
    <h2>
      Edit
       Part Type
    </h2>
    <div>
      <SelectInput
        className=""
        error=""
        fieldName="includeInSection"
        isEmptyAllowed={true}
        label="Section"
        multipleSize={1}
        onChange={[Function]}
        options={
          Array [
            Object {
              "name": "section 1",
              "value": "1",
            },
            Object {
              "name": "section 2",
              "value": "2",
            },
          ]
        }
        title="Select Section"
        value="1"
      />
    </div>
    <PartTypeData
      handlePartTypeValueChange={[Function]}
      partType={
        Object {
          "_detail": true,
          "id": "1",
          "includeInSection": 1,
          "name": "has one one",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "textAlign": "right",
        "width": "100%",
      }
    }
  />
</Modal>
`;
